File: DivideTwoInteger.java
public class DivideTwoInteger {

    //If we left shift the divisor until we cross the value dividend. The n-1 shift required.
    //But there is a change , we left with some remainder so we have to repeat the process again until reminder is less than divisor.

    public int divide(int dividend, int divisor) {
        if (dividend != 0 && divisor == 0)return Integer.MAX_VALUE; //Anything divided by 0 is infinity
       if (dividend == Integer.MIN_VALUE && divisor  == -1) return Integer.MAX_VALUE; //Special test case - ignore

       //Handling overflows
       long ldividend = Math.abs((long) dividend); 
       long ldivisor = Math.abs((long) divisor);
       
       int result = 0;
       while(ldividend >= ldivisor) {
           int noOfShifts = 0;
           while((ldivisor << noOfShifts) <= ldividend) {
               noOfShifts++;    
           }   
           noOfShifts--; //Since it crossed the value of divident, the previous shift was correct.
           result += 1 << noOfShifts;
           ldividend = ldividend - (ldivisor << noOfShifts);
       }
       //If either one of them is negative we should return a negative value.
       if(dividend > 0 && divisor < 0 || dividend <0 && divisor > 0) return -result;
       
       return result;
   }

    public static void main(String[] args) {
        DivideTwoInteger divideTwoInteger = new DivideTwoInteger();
        int result = divideTwoInteger.divide(10, 3);
        System.out.println("The result: " + result);
    }
}

File: DivideTwoInteger.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BitManipulation
{
    /*
     * T.C: O(1) as the loop will at max run for 31 times
     * S.C: O(1) 
     */
    public class DivideTwoInteger
    {
        public int Divide(int dividend, int divisor)
        {
            if (dividend != 0 && divisor == 0) return int.MaxValue;
            if (dividend == int.MinValue && divisor == -1) return int.MaxValue;

            long ldividend = Math.Abs((long)dividend);
            long ldivisor = Math.Abs((long)divisor);
            int result = 0;


            while (ldividend >= ldivisor)
            {

                int shift = 0;
                while (ldividend >= (ldivisor << shift))
                {
                    shift++;
                }
                shift--;
                result = result + (1 << shift);
                ldividend = ldividend - (ldivisor << shift);

            }

            if (dividend > 0 && divisor < 0) return -result;
            if (dividend < 0 && divisor > 0) return -result;

            return result;
        }
    }
}
